### ISSUE TITLE
OpenSSL subprocess times out while going through Xiden Guide

### ISSUE THREAD

**Issue by countvajhula** [zyrolasting/xiden#29]
Hello,
Thank you for your efforts on this package! I was going through the guide and ran into an error on [this step](https://docs.racket-lang.org/xiden-guide/launchers.html#%28part._do%29):

```
$ xi do +a definition.rkt
definition.rkt: fetched
sagegerard.com:my-first-package:default:0, output default
Resolving input "default-local.tgz"

OpenSSL subprocess timed out after #<subprocess> seconds
  exe: /opt/local/bin/openssl
  arguments: ("dgst" "-binary" "-sha3-384")
```

It looks like it might be because of this CLI argument `-sha3-384`. The openssl docs I have locally seem to suggest it would accept `-sha384`. I confirmed that the former is rejected:
```
$ echo "hi" | openssl dgst -binary -sha3-384
unknown option '-sha3-384'
```
while `-sha384` is accepted.

My environment:
Mac OS High Sierra (this is a relatively old version at this point, still, the openssl version seems to be above the minimum mentioned in the xiden docs...)
OpenSSL 1.0.0d 8 Feb 2011
I installed xiden from git, pulled from github earlier today.

**Comment by zyrolasting** [comment]
Thanks for letting me know about this. I has assumed OpenSSL would make different CHFs available depending on the installed version, but it looks like that's not the whole story. For the short term I'll adjust the guide to account for this issue.

**Comment by zyrolasting** [comment]
@countvajhula Can you take a look at #30? I'll merge and close it out in 2 days if I don't hear back, but I want to hear if you think this was handled well. Thanks for raising the issue!

**Comment by countvajhula** [comment]
Hey, sure thing @zyrolasting . I'm trying out those steps and running into
```
car: contract violation
  expected: pair?
  given: '()
```
It looks like this is because in this definition in openssl.rkt, it is attempting to call `car` on `cryptographic-hash-functions` but that value is null.
I just played around with it a bit and looks like

```
(run-openssl-command (open-input-bytes #"")
                                    "list"
                                    "-digest-commands")
```
returns an empty byte string, `#""`. Is that supposed to be
```
(run-openssl-command (open-input-bytes #"")
                                    "list-message-digest-commands")
```
by any chance? That seems to work.

As for how this was handled, you're doing great Sage! One can hardly complain about < 1 day turnaround times 😸 I appreciate your commitment!

**Comment by countvajhula** [comment]
Hmm, now that I'm looking at the Guide, the fact that you've explicitly suggested `openssl list -digest-commands` seems to suggest this is another openssl version disparity issue. Maybe this would be addressed by #31 ?

**Comment by zyrolasting** [comment]
Yep, I think this scenario warrants moving forward on #31. Espectially since I _really_ did not expect anyone to end up with `null` for `cryptographic-hash-functions`!

I'm going to put in a way for you to configure which OpenSSL binary Xiden uses in #30. That way you can at least point it to a separate OpenSSL build that won't encounter these problems while I work on #31 (Obviously be careful not to overwrite the OpenSSL already on your system, because 99% of the time something in the OS really needs it to be exactly the version that it is).

**Comment by countvajhula** [comment]
@zyrolasting No rush on this at all, I'm happy to wait for #31 if it comes to that. That way, I can continue where I left off and validate the fix :)

**Comment by zyrolasting** [comment]
#31 fixed by #34. This issue is no longer in scope because OpenSSL is no longer launched as a subprocess. Please open a new issue for blocking errors. Thank you for the report!
